File: src/magma/Result.java
package magma;

/**
 * A simple result type representing either a successful value or an error.
 *
 * @param <T> the successful value type
 * @param <X> the exception type, which must extend {@link Exception}
 */
public sealed interface Result<T, X extends Exception>
        permits Result.Ok, Result.Err {

    /** Result variant representing success. */
    final class Ok<T, X extends Exception> implements Result<T, X> {
        private final T value;

        public Ok(T value) {
            this.value = value;
        }

        public T value() {
            return value;
        }
    }

    /** Result variant representing failure. */
    final class Err<T, X extends Exception> implements Result<T, X> {
        private final X error;

        public Err(X error) {
            this.error = error;
        }

        public X error() {
            return error;
        }
    }

    /** Factory method for creating a successful result. */
    static <T, X extends Exception> Result<T, X> ok(T value) {
        return new Ok<>(value);
    }

    /** Factory method for creating an error result. */
    static <T, X extends Exception> Result<T, X> err(X error) {
        return new Err<>(error);
    }

    /** Convenience method to check if this result is successful. */
    default boolean isOk() {
        return this instanceof Ok;
    }

    /** Convenience method to check if this result is an error. */
    default boolean isErr() {
        return this instanceof Err;
    }

    /**
     * Gets the successful value or throws the stored exception if this result
     * represents an error.
     */
    default T unwrap() throws X {
        if (this instanceof Ok<T, X> ok) {
            return ok.value();
        }
        throw ((Err<T, X>) this).error();
    }
}

File: src/node/magma/Result.ts
package magma;

/**
 * A generic Result interface representing either a successful operation (Ok) or an error (Err).
 * This is used in place of exceptions for error handling.
 *
 * @param <T> The type of the value in case of success
 * @param <E> The type of the error in case of failure
 */
public interface Result<T, E> {
    /**
     * Checks if this result is an Ok variant.
     *
     * @return true if this is an Ok result, false otherwise
     */
    boolean isOk();

    /**
     * Checks if this result is an Err variant.
     *
     * @return true if this is an Err result, false otherwise
     */
    boolean isErr();

    /**
     * Gets the value if this is an Ok result.
     *
     * @return the value
     * @throws IllegalStateException if this is an Err result
     */
    T getValue();

    /**
     * Gets the error if this is an Err result.
     *
     * @return the error
     * @throws IllegalStateException if this is an Ok result
     */
    E getError();

    /**
     * Creates a new Ok result with the given value.
     *
     * @param value the value
     * @param <T> the type of the value
     * @param <E> the type of the error
     * @return a new Ok result
     */
    static <T, E> Result<T, E> ok(T value) {
        return new Ok<>(value);
    }

    /**
     * Creates a new Err result with the given error.
     *
     * @param error the error
     * @param <T> the type of the value
     * @param <E> the type of the error
     * @return a new Err result
     */
    static <T, E> Result<T, E> err(E error) {
        return new Err<>(error);
    }

    /**
     * An Ok variant of Result representing a successful operation.
     *
     * @param <T> The type of the value
     * @param <E> The type of the error (unused in this variant)
     */
    final class Ok<T, E> implements Result<T, E> {
        private final T value;

        private Ok(T value) {
            this.value = value;
        }

        @Override
isOk() {
            return true;
        }

        @Override
isErr() {
            return false;
        }

        @Override
getValue() {
            return value;
        }

        @Override
getError() {
            throw new IllegalStateException("Cannot get error from Ok result");
        }
    }

    /**
     * An Err variant of Result representing a failed operation.
     *
     * @param <T> The type of the value (unused in this variant)
     * @param <E> The type of the error
     */
    final class Err<T, E> implements Result<T, E> {
        private final E error;

        private Err(E error) {
            this.error = error;
        }

        @Override
isOk() {
            return false;
        }

        @Override
isErr() {
            return true;
        }

        @Override
getValue() {
            throw new IllegalStateException("Cannot get value from Err result");
        }

        @Override
getError() {
            return error;
        }
    }
}